standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership holder MUST be able to extend their membership;
+- any user except the membership holder MUST NOT be able to extend a membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Holding a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
+
+### Withdraw the deposit
+
+Deposit withdrawal is subject to the following conditions:
+- the membership holder MUST be able to withdraw their deposit;
+- any user except the membership holder MUST NOT be able to withdraw its deposit;
+- a deposit MUST be withdrawn in full;
+- a withdrawal MUST fail if the membership is not in _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_;
+- a membership MUST become _Erased_ after withdrawal.
+
+## Governance and upgradability
+
+At initial mainnet deployment, the contract MUST have an _Owner_.
+The _Owner_ MUST be able to change the values of all contract parameters.
+The _Owner_ MUST be able to pause any of the following contract functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+At some point, the _Owner_ SHOULD renounce their privileges, and the contract MUST become immutable.
+Further upgrades, if necessary, SHOULD be done by deploying a new contract and migrating the membership set.
+
+## Implementation Suggestions
+
+User-facing application SHOULD suggest a few rate limits (tiers) to simplify their users' choice.
+The RECOMMENDED rate limits in a three-tier model are as follows:
+- `20` messages per epoch as low-tier;
+- `200` messages per epoch as mid-tier;
+- `600` messages per epoch as high-tier.
+Users SHOULD be able to select a custom rate limit under advanced settings.
+
+The RECOMMENDED pricing parameters are:
+
+| Parameter          | Value  |
+| ------------------ | ------ |
+| Accepted tokens    | `DAI`  |
+| Reference currency | `USD`  |
+| Pricing function   | linear |
+
+User-facing applications SHOULD save membership expiration dates in a local keystore during registration,
+and notify the user when their membership is about to expire.
+
+## Q&A
+
+### Why can't I withdraw a deposit from an _Active_ membership?
+
+The rationale for this limitation is to prevent an undesirable usage pattern where users make deposits and withdrawals in short succession.
+
+### Why can't I extend an _Active_ membership?
+
+Memberships can only be extended during _GracePeriod_.
+We do not allow extending an _Active_ membership.
+The rationale is that if the contract _Owner_ changes some contract parameters (e.g., for security purposes),
+users with extended memberships will not be affected by the changes for a long time.
+
+### What if I don't extend my membership within its _GracePeriod_?
+
+The user who does not extend their _GracePeriod_ membership, 
+assume the risk of the membership being overwritten at any moment.
```

Comment:
```suggestion
assume the risk of the membership being overwritten as soon as it _expires_.
```

Reply:
If a membership is past its grace period, it is by definition in the _Expired_ state. For me it would sound confusing if an (already) _Expired_ membership can _expire_ (this time for real).

The core reason of this confusion is that we have two ways to look at a membership:
- what is its state from the protocol perspective;
- what is its state as reflected in contract storage.

The state diagram follows the former definition.

After the grace period expires, the membership becomes "logically" _Expired_, but it's not immediately reflected in the contract, hence (quite paradoxically) an _Expired_ membership allows sending messages until it becomes _Erased_ from the tree (or _ErasedAwaitsWithdrawal_).

I'd suggest re-formulating to:

> assume the risk of their _Expired_ membership being overwritten at any moment.

or simply

> 
> assume the risk of their _Expired_ membership being overwritten.